### PR TITLE
Issue #2831: Support WINDOW FILTER

### PR DESCRIPTION
**Pull request by hawkfish**
Add support for filtering windowed aggregates.
Extend the window API to accept the filter mask.

### CODE REVIEW

File: src/storage/storage_info.cpp [hunk 1]
```diff
@@ -2,6 +2,6 @@
 
 namespace duckdb {
 
-const uint64_t VERSION_NUMBER = 33;
+const uint64_t VERSION_NUMBER = 34;
```
Comment (Mytherin) [inline src/storage/storage_info.cpp:5]: Can we do this without increasing the storage version?

File: src/parser/expression/window_expression.cpp [hunk 2]
```diff
@@ -120,6 +127,7 @@ void WindowExpression::Serialize(FieldWriter &writer) const {
 	for (auto &order : orders) {
 		order.Serialize(serializer);
 	}
+	writer.WriteOptional(filter_expr);
```
Comment (Mytherin) [inline src/parser/expression/window_expression.cpp:130]: In order to not increase the storage version, this needs to be written as the last element in the window expression

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Some comments relating to the storage:

File: src/parser/expression/window_expression.cpp (hunk 2 above)
Comment (hawkfish) [reply src/parser/expression/window_expression.cpp:130]: Sneaky!  I'll give that a whirl. 

I'm not clear on why this works, though? Doesn't optional mean you write a flag?

Comment (Mytherin) [reply src/parser/expression/window_expression.cpp:130]: Optional writes a boolean followed by the actual element if the boolean is true, but the actual bool must still be read. The `FieldWriter` allows us to skip fields if they are not present in the file by looking at the field count that is written as a header, see #2969. For this to work all new elements must be added at the end, however.

Comment (hawkfish) [reply src/parser/expression/window_expression.cpp:130]: Ah, that makes sense. And it worked!

**Comment by Mytherin** [comment]
LGTM

@lnkuiper could you have a last look?

**Review by lnkuiper — APPROVED** [review]
Looks good to me! Clean implementation too, good stuff Richard!

**Pull request merged**